blockchain/consensus/narwhal_cns/src/proposer.rs

```diff
@@ -0,0 +1,420 @@
+        let state_manager_clone = state_manager.clone();
+        handles.push(task::spawn(async move {
+            if let Err(e) = handle_head_changes(
+                state_manager_clone,
+                submitter,
+                self.validator_addresses,
+                current_head,
+                head_change_rx,
+                head_tx,
+                output_rx,
+            )
+            .await
+            {
+                error!("Error handling head changes: {}", e)
+            }
+        }));
+
+        handles.push(task::spawn(async move {
+            if let Err(e) =
+                handle_head_watch(state_manager, mpool, head_rx, transactions_client).await
+            {
+                error!("Error handling head watch: {}", e)
+            }
+        }));
+
+        Ok(handles)
+    }
+}
+
+fn tokio_to_async_std(
+    handles: Vec<tokio::task::JoinHandle<()>>,
+) -> Vec<async_std::task::JoinHandle<()>> {
+    handles
+        .into_iter()
+        .map(|h| task::spawn(async { h.await.expect("The tokio task has panicked.") }))
+        .collect()
+}
+
+/// Handle local chain extensions by projecting blocks on top of them from the Narwhal batches.
+/// Also ping the mempool reader process that there might be some change that makes it worth running another check.
+async fn handle_head_changes<DB>(
+    state_manager: Arc<StateManager<DB>>,
+    submitter: SyncGossipSubmitter,
+    validator_addresses: HashMap<PublicKey, Address>,
+    mut current_head: Option<Arc<Tipset>>,
+    mut head_change_rx: tokio::sync::broadcast::Receiver<HeadChange>,
+    head_tx: tokio::sync::watch::Sender<Option<Arc<Tipset>>>,
+    output_rx: Receiver<NarwhalOutput>,
+) -> anyhow::Result<()>
+where
+    DB: BlockStore + Sync + Send + 'static,
+{
+    loop {
+        // Wait for the next chain extension.
+        let next_head = if let Some(tipset) = current_head.take() {
+            tipset
+        } else {
+            match head_change_rx.recv().await {
+                Err(_) => {
+                    return Err(anyhow!("Cannot receive head changes!"));
+                }
+                Ok(HeadChange::Current(_)) => panic!("Did not expect to receive the current head."),
+                Ok(HeadChange::Revert(_)) => {
+                    panic!("This consensus is supposed to be forward only!")
+                }
+                Ok(HeadChange::Apply(tipset)) => tipset,
+            }
+        };
+
+        // Notify the mempool polling process that there's a new tip.
+        head_tx.send_replace(Some(next_head.clone()));
+
+        // Take the next certificate from the queue and turn it into a block, then submit.
+        let next_output = match output_rx.recv().await {
+            Err(_) => return Err(anyhow!("Cannot receive Narwhal output!")),
+            Ok(output) => output,
+        };
+
+        // TODO: We might have to create multiple blocks from the same batch to respect limits!
```

Comment:
You mean so our blocks don't overflow the gas limit? This is something that we should probably consider making configurable, I don't know if gas limits make sense for steam-like consensus like Narwhal

Reply:
That's a good point, if the purpose of the gas limit is to avoid blocks being so large that they can't be propagated across the network in a timely manner, then with this deterministic derivation it doesn't really matter. Currently there are [hardcoded limits](https://github.com/aakoshh/forest/blob/main/blockchain/chain_sync/src/tipset_syncer.rs#L1446), and I didn't know if there are other repercussions of changing this, as it's coming from the `fvm_shared` package.

Reply:
I don't think there is an internal check by the FVM to see if the gas limit of the block is reached. This is only enforced by the peer as part of the consensus validation. The FVM only checks that the gas spent by a message is not over the gas limit for that specific message.

Reply:
It turns out that I _must_ disable gas limits because: 
1. I realised that there while there is no notion of a timestamp in the FVM, contracts do receive `epoch`, which they can use for time locks for example, but only if it can be expected to follow some regularity. The way I turned each certificate into its own epoch meant that there could be more or less epochs depending on committee size and network circumstances (how many got quorum). That would be unpredictable for smart contracts.
2. I decided that I will group every certificate that appears in the same Narwhal round into a single Filecoin epoch. Narwhal rounds have a timeout, so we at least have a notion of what to expect. This means each Narwhal certificate within the round will go in the same Tipset.
3. The Filecoin Tipset requires each miner in it to be unique.
4. Therefore I cannot split up a batch between blocks and put them in the same Tipset, because the miner would be the same.